### PR TITLE
ARROW-3050: [C++] Adopt HiveServer2 client codebase from cloudera/hs2client. Add Thrift to thirdparty toolchain

### PR DESCRIPTION
**Pull request by wesm**
I have forked the C++ part of http://github.com/cloudera/hs2client (which I collaborated on some time ago) and modified it to work with the Arrow CMake build system. It can be built with `-DARROW_HIVESERVER2=ON`. This also adds Thrift to our thirdparty toolchain as an optional EP (the configuration was lifted from parquet-cpp)

Open questions

- [x] Make sure unit tests work
- [x] Attribute IP lineage in LICENSE.txt
- I have included the extensions to HS2 that are available in Impala, so this should work with both Hive and Impala

Follow up work:

- Dockerized testing setup (it wouldn't really make sense to test this in Travis CI)
- Nightly testing of docker tests
- Convert HS2 columnar row sets to Arrow record batches
- Add Python bindings
- Add DBABI wrapper for Python
- Add SASL support for Kerberos-enabled HS2
- Add SSL socket support
- Add LDAP PLAIN authentication support

### CODE REVIEW

File: cpp/src/arrow/status.h [hunk 1]
```diff
@@ -84,6 +84,7 @@ enum class StatusCode : char {
   PlasmaObjectNonexistent = 21,
   PlasmaStoreFull = 22,
   PlasmaObjectAlreadySealed = 23,
+  StillExecuting = 24
```
Comment (wesm) [inline cpp/src/arrow/status.h]: Wasn't sure about this. We might call this a more generic "Not ready" status code

**Comment by wesm** [comment]
I'm also thinking whether we want to create a new `arrow/dbi` subdirectory rather than `adapters` since we may implement other Arrow native DB interfaces in the future (e.g. PostgreSQL would be of interest to me)

**Comment by wesm** [comment]
The unit tests pass fine when you have Impala running on localhost. I added a docker-compose setup but wasn't able to get the dependent Impala container to start up properly. If @cpcloud, @xhochy, or @kszucs or another Docker expert could take a look it would be much appreciated

**Comment by wesm** [comment]
I know that this PR is big. I'd like to proceed with the follow up work and would prefer to do so in smaller PRs. Let me know if you have any strong opinions about what you see so far

**Comment by pcmoritz** [comment]
No objection to this PR, but I wonder if longer term we can standardize on one RPC framework for arrow. This would make it easier to tie together the various services that live in the repo. My personal preference would be gRPC (but this is based off limited experience) and I'd be interested in exposing plasma as a gRPC service.

**Comment by xhochy** [comment]
I'm a bit -1 on merging it directly merging it. As far as I could see from a quick glance, it currently only uses Arrow's CMake infrastructure and Status infrastructure. 

@wesm some questions (feel free to answer them on the ML)

- Is the intention to then discontinue the hs2client project and instead evolve everything here?
- Would you also try to get rid of some of the hs2client classes in favour of the similar ones in Arrow?
- We have some DBABI extensions in `turbodbc` for retrieving and submitting Arrow. Would it make sense to propose a PEP that extends the DBAPI spec so that they are standardized?

**Comment by wesm** [comment]
See related discussion on cloudera/hs2client https://github.com/cloudera/hs2client/issues/27

**Comment by xhochy** [comment]
+1, feel free to merge then.

**Pull request closed without merging**